grove/deutsch_jozsa/deutsch_jozsa.py

```diff
@@ -0,0 +1,159 @@
+
+        # Apply Hadamard, Oracle, and Hadamard again
+
+        dj_prog.inst([H(qubit) for qubit in self.computational_qubits])
+
+        # Build the oracle
+        oracle_prog = pq.Program()
+        oracle_prog.defgate(ORACLE_GATE_NAME, self.unitary_matrix)
+
+        scratch_bit = self.ancillas[1]
+        qubits_for_funct = [scratch_bit] + self.computational_qubits
+        oracle_prog.inst(tuple([ORACLE_GATE_NAME] + qubits_for_funct))
+        dj_prog += oracle_prog
+
+        # Here the oracle does not leave the computational qubits unchanged, so we use a CNOT to
+        # to move the result to the query qubit, and then we uncompute with the dagger.
+        dj_prog.inst(CNOT(self._qubits[0], self.ancillas[0]))
+        dj_prog += oracle_prog.dagger()
+        dj_prog.inst([H(qubit) for qubit in self.computational_qubits])
+        return dj_prog
+
+    @staticmethod
+    def unitary_function(mappings):
+        """
+        Creates a unitary transformation that maps each state to the values specified
+        in mappings.
+
+        Some (but not all) of these transformations involve a scratch qubit, so room for one is
+        always provided. That is, if given the mapping of n qubits, the calculated transformation
+        will be on n + 1 qubits, where the 0th is the scratch bit and the return value
+        of the function is left in the 1st.
+
+        :param mappings: Dictionary of the mappings of f(x) on all length n bitstrings, e.g.
+
+            >>> {'00': '0', '01': '1', '10': '1', '11': '0'}
+
+        :type mappings: Dict[String, Int]
+        :return: ndarray representing specified unitary transformation.
+        :rtype: np.ndarray
+        """
+        num_qubits = int(np.log2(len(mappings)))
+        bitsum = sum([int(bit) for bit in mappings.values()])
+
+        # Only zeros were entered
+        if bitsum == 0:
+            return np.kron(SWAP_MATRIX, np.identity(2 ** (num_qubits - 1)))
+
+        # Half of the entries were 0, half 1
+        elif bitsum == 2 ** (num_qubits - 1):
+            unitary_funct = np.zeros(shape=(2 ** num_qubits, 2 ** num_qubits))
+            index_lists = [list(range(2 ** (num_qubits - 1))),
+                           list(range(2 ** (num_qubits - 1), 2 ** num_qubits))]
+            for j in range(2 ** num_qubits):
+                bitstring = bin(j)[2:]
+                value = int(mappings[bitstring])
+                mappings.pop(bitstring)
+                i = index_lists[value].pop()
+                unitary_funct[i, j] = 1
+            return np.kron(np.identity(2), unitary_funct)
+
+        # Only ones were entered
+        elif bitsum == 2 ** num_qubits:
+            x_gate = np.array([[0, 1], [1, 0]])
+            return np.kron(SWAP_MATRIX, np.identity(2 ** (num_qubits - 1))).dot(
+                np.kron(x_gate, np.identity(2 ** num_qubits)))
+        else:
+            raise ValueError("f(x) must be constant or balanced")
```

Comment:
You just wrapped the classic algorithm to check the solution to generate the oracle. This renders the quantum algorithm fairly useless? Am I missing something?

Reply:
Can you elaborate on what you mean? The function is provided as a bitmap, and the values are iterated over, and the matrix is built. What do you mean by "check the solution to generate the oracle"?

Reply:
Well, what I mean by this, is that to generate the oracle you walk over the full bitmap, so you can find the solution by just checking every element.

Reply:
I'll address this in documentation.